src/com/owncloud/android/ui/activity/Uploader.java

```diff
@@ -520,110 +474,129 @@ private boolean somethingToUpload() {
+                        requestUpload(sourceUri.getPath(), remotePath);
+                        schemeFileCounter++;
                     }
-                    //if (data == null) {
-                        mRemoteCacheData.add(filePath);
-                        CopyTmpFileAsyncTask copyTask = new CopyTmpFileAsyncTask(this);
-                        Object[] params = {uri, filePath, mRemoteCacheData.size() - 1,
-                                getAccount().name, getContentResolver()};
-                        mNumCacheFile++;
-                        showWaitingCopyDialog();
-                        copyTask.execute(params);
-
-                    //} else {
-                    // TODO request to FileUploader with data as source file, resulting in lazy temporary copy
-                    //}
-                } else {
-                    throw new SecurityException();
                 }
+            }
+
+            if (!contentUris.isEmpty()) {
+                /// content: uris will be copied to temporary files before calling {@link FileUploader}
+                copyThenUpload(contentUris.toArray(new Uri[contentUris.size()]),
+                    contentRemotePaths.toArray(new String[contentRemotePaths.size()]));
 
-                //Save the path to shared preferences
-                SharedPreferences.Editor appPrefs = PreferenceManager
-                        .getDefaultSharedPreferences(getApplicationContext()).edit();
-                appPrefs.putString("last_upload_path", mUploadPath);
-                appPrefs.apply();
+            } else if (schemeFileCounter == 0) {
+                showErrorDialog(
+                    R.string.uploader_error_message_no_file_to_upload,
+                    R.string.uploader_error_title_no_file_to_upload
+                );
 
+            } else {
                 finish();
             }
 
         } catch (SecurityException e) {
-            String message = String.format(getString(R.string.uploader_error_forbidden_content),
-                    getString(R.string.app_name));
-            Toast.makeText(this, message, Toast.LENGTH_LONG).show();
+            Log_OC.e(TAG, "Permissions fail", e);
+            showErrorDialog(
+                R.string.uploader_error_message_read_permission_not_granted,
+                R.string.uploader_error_title_file_cannot_be_uploaded
+            );
+
+        } catch (Exception e) {
+            Log_OC.e(TAG, "Unexpted error", e);
+            showErrorDialog(
+                R.string.common_error_unknown,
+                R.string.uploader_error_title_file_cannot_be_uploaded
+            );
+
+        } finally {
+            // Save the path to shared preferences; even if upload is not possible, user chose the folder
+            SharedPreferences.Editor appPrefs = PreferenceManager
+                .getDefaultSharedPreferences(getApplicationContext()).edit();
+            appPrefs.putString("last_upload_path", mUploadPath);
```

Comment:
"last_upload_path" should be a constant, so someone how wants to access it can also refer to the same constant :)